### PR TITLE
Improve push notification registration

### PR DESCRIPTION
**Pull request by vargaat**
### What changed?
- There was an issue with push notifications that caused the app not to register the user for push notifications after logging out and logging in with the same user again. This was because the shared instance's internal state was not properly reset during logout.
- Since the NotificationsManager was quite bloated I split this into smaller files, the two most important are: LocalNotificationsInteractor and PushNotificationsInteractor. Most of the code changes are because of this refactor.
- I also grouped files into folders based on their responsibility in the Notifications Xcode group.
- Changed method names related to push notifications for clarity.
- Added meaningful unit tests for push notification subscriptions.
- Fixed an issue that caused API mocks for the same path but with different HTTP methods to collide.

refs: [MBL-17418](https://instructure.atlassian.net/browse/MBL-17418)
affects: Student, Teacher
release note: Fixed push notifications not arriving after logging out and logging back in with the same user.

test plan: See ticket.

## Checklist

- [ ] Follow-up e2e test ticket created
- [x] Tested on phone
- [ ] Tested on tablet

[MBL-17418]: https://instructure.atlassian.net/browse/MBL-17418?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>



#### Release Note: 
Fixed push notifications not arriving after logging out and logging back in with the same user.

#### Affected Apps: Student, Teacher

[MBL-17418](https://instructure.atlassian.net/browse/MBL-17418)

<details><summary>:x: Swift lint</summary>


    ❌ /Users/vagrant/git/scripts/swift/SummarizeTestResults.swift:88:19: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/scripts/swift/SummarizeTestResults.swift:107:24: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/scripts/swift/main.swift:33:36: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncModulesInteractorLiveTests.swift:92:23: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncFilesInteractorLiveTests.swift:69:90: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncFilesInteractorLiveTests.swift:93:23: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncFilesInteractorLiveTests.swift:140:23: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncFilesInteractorLiveTests.swift:258:23: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/CourseSync/CourseSyncDownloader/Model/Downloaders/CourseSyncFilesInteractorLiveTests.swift:262:23: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Assignments/APIAssignmentListTests.swift:59:23: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/UploadMediaCommentTests.swift:75:76: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/UploadMediaCommentTests.swift:77:85: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/AudioVideo/UploadMediaCommentTests.swift:101:76: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/UploadMediaCommentTests.swift:103:85: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/APIMediaCommentTests.swift:26:74: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/AudioVideo/APIMediaCommentTests.swift:52:13: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/APIMediaCommentTests.swift:56:13: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AudioVideo/APIMediaCommentTests.swift:71:89: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/PageViewAnalytics/APIPandataTests.swift:59:43: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/PageViewAnalytics/PageViewEventRequestManagerTests.swift:75:95: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/PageViewAnalytics/PageViewEventRequestManagerTests.swift:94:38: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/PageViewAnalytics/PageViewEventRequestManagerTests.swift:97:84: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Extensions/URLResponseExtensionsTests.swift:51:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Extensions/URLResponseExtensionsTests.swift:64:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Extensions/URLResponseExtensionsTests.swift:71:20: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Extensions/FileManagerExtensionsTests.swift:31:64: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Conversations/APIConversationTests.swift:30:106: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Files/Model/API/APIFileTests.swift:103:20: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/View/FilePicker/FilePickerViewControllerTests.swift:60:82: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/PrepareUpload/FileSubmissionComposerTests.swift:27:76: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Files/View/FileDetails/FileDetailsViewControllerTests.swift:110:69: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/View/FileDetails/FileDetailsViewControllerTests.swift:114:52: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/View/FileDetails/FileDetailsViewControllerTests.swift:119:69: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Files/View/FileDetails/FileDetailsViewControllerTests.swift:123:52: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/BinaryUpload/BackgroundURLSessionProviderTests.swift:83:75: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/BinaryUpload/BackgroundURLSessionProviderTests.swift:84:71: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/PrepareUpload/FileUploadTargetRequesterTests.swift:28:76: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/FileSubmissionAssemblyTests.swift:145:68: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/FileSubmission/Model/Submission/FileSubmissionCleanupTests.swift:28:76: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/API/TypeSafeCodableTests.swift:70:24: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/TypeSafeCodableTests.swift:75:24: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/TypeSafeCodableTests.swift:80:24: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/API/APIFormDataTests.swift:36:20: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/APIFormDataTests.swift:38:22: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/APIFormDataTests.swift:54:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/API/APIFormDataTests.swift:63:28: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/IDTests.swift:42:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/SafeURLTests.swift:119:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Files/UploadManagerTests.swift:39:68: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Files/UploadManagerTests.swift:83:68: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/APITests.swift:55:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/API/APITests.swift:368:29: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/APIRequestableTests.swift:130:88: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/API/APIRequestableTests.swift:204:29: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/API/APIRequestableTests.swift:213:29: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/DocViewer/APIDocViewerTests.swift:39:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/DocViewer/APIDocViewerTests.swift:44:21: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/DocViewer/APIDocViewerTests.swift:75:20: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/DocViewer/APIDocViewerTests.swift:80:21: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AppEnvironment/CacheManagerTests.swift:91:14: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/AppEnvironment/KeychainTests.swift:40:22: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AppEnvironment/KeychainTests.swift:48:18: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/AppEnvironment/KeychainTests.swift:76:25: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Offline/Model/OfflineFileInteractorTests.swift:40:23: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Offline/Model/OfflineFileInteractorTests.swift:57:23: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/CoreTests/Offline/Model/OfflineFileInteractorTests.swift:76:23: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/CoreTests/Offline/Model/OfflineFileInteractorTests.swift:93:23: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Parent/Parent/ParentAppDelegate.swift:332:81: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    /Users/vagrant/git/Core/Core/Discussions/DiscussionDetailsViewController.swift:468:9: error: Opening Brace Spacing Violation: Opening braces should be preceded by a single space and on the same line as the declaration (opening_brace)
    ❌ /Users/vagrant/git/Core/Core/PageViewAnalytics/APIPandata.swift:136:16: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/Core/PageViewAnalytics/PageViewEventController.swift:211:16: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/Core/DeveloperMenu/PushNotificationDebugView.swift:32:24: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/Core/Extensions/URLResponseExtensions.swift:57:30: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/Core/LTI/EmbeddedExternalTools.swift:25:39: error: Superfluous Disable Command Violation: SwiftLint rule 'opening_brace' did not trigger a violation in the disabled region; remove the disable command (superfluous_disable_command)
    /Users/vagrant/git/Core/Core/UITestHelpers/IPC.swift:72:34: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/Core/Files/View/FileDetails/FileDetailsViewController.swift:562:107: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    /Users/vagrant/git/Core/Core/API/APIMock.swift:151:38: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/Core/API/API.swift:76:52: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/Core/RichContentEditor/RichContentEditorViewController.swift:176:25: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/Core/RichContentEditor/RichContentEditorViewController.swift:380:35: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/rn/Teacher/ios/Teacher/TeacherAppDelegate.swift:474:81: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    ❌ /Users/vagrant/git/Core/TestsFoundation/MiniCanvas/LoggingHttpServer.swift:60:38: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:188:30: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:193:30: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:205:37: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:298:36: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:498:19: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/TestsFoundation/UITests/CoreUITestCase.swift:508:35: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/MiniCanvas/State/MiniCourse.swift:174:24: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Core/TestsFoundation/MiniCanvas/MiniCanvasEndpoints.swift:132:19: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Core/TestsFoundation/Helpers/CommandLineHelper.swift:81:50: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Student/StudentUnitTests/Assignments/AssignmentDetails/AssignmentDetailsPresenterTests.swift:518:68: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Student/StudentUnitTests/KeychainTests.swift:31:21: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/Student/StudentUnitTests/Submissions/SubmissionButton/SubmissionButtonPresenterTests.swift:265:68: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/rn/Teacher/ios/TeacherTests/Attendance/RollCallSessionTests.swift:80:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/rn/Teacher/ios/TeacherTests/Attendance/RollCallSessionTests.swift:93:20: error: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    ❌ /Users/vagrant/git/rn/Teacher/ios/TeacherTests/KeychainTests.swift:31:21: Non-Optional String <-> Data Conversion Violation: Prefer using UTF-8 encoded strings when converting between `String` and `Data` (non_optional_string_data_conversion)
    /Users/vagrant/git/Student/Student/Submissions/SubmissionButton/SubmissionButtonPresenter.swift:313:55: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    /Users/vagrant/git/Student/Student/Submissions/SubmissionButton/SubmissionButtonPresenter.swift:329:58: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    ❌ /Users/vagrant/git/Student/Student/Submissions/SubmissionButton/SubmissionButtonPresenter.swift:335:67: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    /Users/vagrant/git/Student/Student/StudentAppDelegate.swift:507:81: error: Redundant Void Return Violation: Returning Void in a function declaration is redundant (redundant_void_return)
    /Users/vagrant/git/Student/Widgets/AnnouncementsWidget/Model/GetWidgetAnnouncements.swift:49:19: error: Unused Enumerated Violation: When the index is not used, `.enumerated()` can be removed (unused_enumerated)
</details>
<details><summary>:x: XCTest failed: CoreTests/DiscussionDetailsViewControllerTests/testLayout</summary>


    XCTAssertTrue failed
    XCTAssertEqual failed: ("Optional("Mark as Unread")") is not equal to ("Optional("Mark as Read")")
</details>
<details><summary>:x: XCTest failed: CoreTests/FileAccessReportInteractorTests/testFinishesOnSuccessfulReport</summary>


    failed - Unexpected failure while waiting on finish event: Error Domain=com.instructure Code=0 "No response or error received." UserInfo={NSLocalizedDescription=No response or error received.}
</details>
<details><summary>:x: XCTest failed: CoreTests/FileSubmissionItemsUploadStarterTests/testStartsFileUploadAndSetsTaskIDOnUploadTask</summary>


    failed - Upload session not started.
</details>
<details><summary>:x: XCTest failed: TeacherTests/AttendanceStatusControllerTests/testUpdate</summary>


    API violation - multiple calls made to -[XCTestExpectation fulfill] for errored. (NSInternalInconsistencyException)
    Failed due to unwaited expectation 'changed'.
</details>
<details><summary>:x: XCTest failed: TeacherTests/AttendanceViewControllerTests/testChangeDate</summary>


    API violation - multiple calls made to -[XCTestExpectation fulfill] for errored. (NSInternalInconsistencyException)
</details>
<details><summary>:x: XCTest failed: TeacherTests/RollCallSessionTests/testUpdateStatus</summary>


    XCTAssertNil failed: "Error Domain=com.instructure.rollcall Code=1 "Error: No data returned from the rollcall api." UserInfo={NSLocalizedDescription=Error: No data returned from the rollcall api.}"
    XCTAssertEqual failed: ("nil") is not equal to ("Optional("1")")
    XCTAssertNil failed: "Error Domain=com.instructure.rollcall Code=1 "Error: No data returned from the rollcall api." UserInfo={NSLocalizedDescription=Error: No data returned from the rollcall api.}"
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 45bdf3bad49e487351cfb5b7ac246bc1b359095b
</p>

**Comment by szabinst** [comment]
I was actually never been able to achieve the repro steps with the store build because `DELETE communication_channels` always fails. And the `POST` counterpart fails too. Same with this PR's build so I can't verify any improvement.

File: Core/Core/Notifications/LocalNotificationsInteractor.swift [hunk 1]
```diff
@@ -73,4 +80,70 @@ public extension NotificationManager {
...
+        let request = UNNotificationRequest(
+            identifier: identifier,
+            title: title,
+            body: body,
+            route: route
+        )
+        notificationCenter.add(request) { error in
+            if let error = error {
+                Analytics.shared.logError(name: "Failed to schedule local notification",
+                                          reason: error.localizedDescription)
```
Comment (szabinst) [inline Core/Core/Notifications/LocalNotificationsInteractor.swift:102]: Food for thought, we could rethink (in the future) sending localized strings to logger services. It segments the data too much imo.

**Review by szabinst — APPROVED** [review]
QA + 1

**Review by ndrsszsz — APPROVED** [review]
QA +1

**Pull request merged**